### PR TITLE
Support request level part size configuration

### PR DESCRIPTION
**Pull request by TingDaoK**
- Add support for request level override for part size and upload threshold
- Add support for on_done callback with the did_validate checksum and the checksum algorithm.
- Submodules update for several S3 related improvements:
  - Fix Get-Object with partNumber
  - Skip HeaderRequest for ranged get
  - Cancel/Pause s3 request now faster (will cancel the ongoing HTTP requests now)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/test_s3.py [hunk 1]
```diff
@@ -260,6 +260,8 @@ def setUp(self):
         self.done_error_headers = None
         self.done_error_body = None
         self.done_error_operation_name = None
+        self.did_validate_checksum = None
+        self.checksum_validation_algorithm = None
```
Comment (graebm) [inline test/test_s3.py:264]: naming pattern is `self.done_XYZ` for stuff that comes from the on_done callback

File: test/test_s3.py [hunk 2]
```diff
@@ -305,12 +307,23 @@ def _on_request_headers(self, status_code, headers, **kargs):
...
+            did_validate_checksum,
+            checksum_validation_algorithm,
+            **kwargs):
         self.done_error = error
         self.done_error_headers = error_headers
         self.done_error_body = error_body
         self.done_error_operation_name = error_operation_name
         self.done_status_code = status_code
+        self.did_validate_checksum = did_validate_checksum
+        self.checksum_validation_algorithm = checksum_validation_algorithm
```
Comment (graebm) [inline test/test_s3.py:326]: down in `_validate_successful_response()` do:
```
self.assertIsNotNone(self.done_did_validate_checksum)

if self.done_did_validate_checksum:
    self.assertIsInstance(self.done_checksum_validation_algorithm, S3ChecksumAlgorithm)
```

File: awscrt/s3.py [hunk 3]
```diff
@@ -657,7 +702,9 @@ def _on_finish(self, error_code, status_code, error_headers, error_body, error_o
                 error_headers=error_headers,
                 error_body=error_body,
                 error_operation_name=error_operation_name,
-                status_code=status_code)
+                status_code=status_code,
+                did_validate_checksum=did_validate_checksum,
+                checksum_validation_algorithm=checksum_validation_algorithm)
```
Comment (graebm) [inline awscrt/s3.py:707]: you're passing `checksum_validation_algorithm` as an int right now, you need to change it into `Optional[S3ChecksumAlgorithm]`

File: awscrt/s3.py [hunk 4]
```diff
@@ -401,6 +418,13 @@ def make_request(
                         this is the final response's status code. If the operation
                         failed for another reason, None is returned.
 
+                    *   `did_validate_checksum` (Optional[bool]):
```
Comment (graebm) [inline awscrt/s3.py:421]: it's not optional, it's always true or false
```suggestion
                    *   `did_validate_checksum` (bool):
```

File: awscrt/s3.py [hunk 5]
```diff
@@ -696,6 +695,11 @@ def _on_finish(
             self._finished_future.set_exception(error)
         else:
             self._finished_future.set_result(None)
+        if checksum_validation_algorithm:
```
Comment (graebm) [inline awscrt/s3.py:698]: utterly trivial: newline between previous if/else block and new if/else block
```suggestion

        if checksum_validation_algorithm:
```

File: test/test_s3.py [hunk 6]
```diff
@@ -336,6 +336,11 @@ def _validate_successful_response(self, is_put_object):
         self.assertIsNone(self.done_error_headers)
         self.assertIsNone(self.done_error_body)
         self.assertIsNone(self.done_error_operation_name)
+        self.assertIsNotNone(self.done_did_validate_checksum)
```
Comment (graebm) [inline test/test_s3.py:339]: trivial:
```suggestion
        self.assertIsInstance(self.done_did_validate_checksum, bool)
```

File: test/test_s3.py [hunk 7]
```diff
@@ -604,8 +609,7 @@ def test_put_get_with_checksum(self):
         download_checksum_config = S3ChecksumConfig(validate_response=True)
         self._test_s3_put_get_object(download_request, S3RequestType.GET_OBJECT,
                                      checksum_config=download_checksum_config)
-        self.assertTrue(self.did_validate_checksum)
-        self.assertEqual(self.checksum_validation_algorithm, S3ChecksumAlgorithm.CRC32)
+        self.assertTrue(self.done_did_validate_checksum)
         self.assertEqual(HttpHeaders(self.response_headers).get('x-amz-checksum-crc32'),
```
Comment (graebm) [inline test/test_s3.py:613]: why did you take this line out? don't we want to validate that the correct checksum type carried through?
```suggestion
        self.assertTrue(self.done_did_validate_checksum)
        self.assertEqual(self.done_checksum_validation_algorithm, S3ChecksumAlgorithm.CRC32)
        self.assertEqual(HttpHeaders(self.response_headers).get('x-amz-checksum-crc32'),
```

**Review by graebm — APPROVED** [review]
fix & ship

File: test/test_s3.py (hunk 7 above)
Comment (TingDaoK) [reply test/test_s3.py]: Ops, I thought it was checked from the `_validate_successful_response`, turns out we are not invoking it from this test.

Comment (graebm) [reply test/test_s3.py]: _validate_successful_response just checks that ANY checksum was set, here we can ensure it was exactly the one we expected

**Pull request merged**